File: go/test/render_page_post_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/phantomjscloud-sdk"
	"github.com/voxgig-sdk/phantomjscloud-sdk/core"

	vs "github.com/voxgig/struct"
)

func TestRenderPagePostEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.RenderPagePost(nil)
		if ent == nil {
			t.Fatal("expected non-nil RenderPagePostEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := render_page_postBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "render_page_post." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set PHANTOMJSCLOUD_TEST_RENDER_PAGE_POST_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		renderPagePostRef01Ent := client.RenderPagePost(nil)
		renderPagePostRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "render_page_post"}, setup.data), "render_page_post_ref01"))
		renderPagePostRef01Data["api_key"] = setup.idmap["api_key01"]

		renderPagePostRef01DataResult, err := renderPagePostRef01Ent.Create(renderPagePostRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		renderPagePostRef01Data = core.ToMapAny(renderPagePostRef01DataResult)
		if renderPagePostRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}

	})
}

func render_page_postBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "render_page_post", "RenderPagePostTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read render_page_post test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse render_page_post test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"render_page_post01", "render_page_post02", "render_page_post03", "api_key01"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("PHANTOMJSCLOUD_TEST_RENDER_PAGE_POST_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"PHANTOMJSCLOUD_TEST_RENDER_PAGE_POST_ENTID": idmap,
		"PHANTOMJSCLOUD_TEST_LIVE":      "FALSE",
		"PHANTOMJSCLOUD_TEST_EXPLAIN":   "FALSE",
		"PHANTOMJSCLOUD_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["PHANTOMJSCLOUD_TEST_RENDER_PAGE_POST_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["PHANTOMJSCLOUD_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["PHANTOMJSCLOUD_APIKEY"],
			},
			extra,
		})
		client = sdk.NewPhantomjscloudSDK(core.ToMapAny(mergedOpts))
	}

	live := env["PHANTOMJSCLOUD_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["PHANTOMJSCLOUD_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: rb/test/render_page_post_entity_test.rb
# RenderPagePost entity test

require "minitest/autorun"
require "json"
require_relative "../Phantomjscloud_sdk"
require_relative "runner"

class RenderPagePostEntityTest < Minitest::Test
  def test_create_instance
    testsdk = PhantomjscloudSDK.test(nil, nil)
    ent = testsdk.RenderPagePost(nil)
    assert !ent.nil?
  end

  def test_basic_flow
    setup = render_page_post_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["create"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "render_page_post." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set PHANTOMJSCLOUD_TEST_RENDER_PAGE_POST_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # CREATE
    render_page_post_ref01_ent = client.RenderPagePost(nil)
    render_page_post_ref01_data = Helpers.to_map(Vs.getprop(
      Vs.getpath(setup[:data], "new.render_page_post"), "render_page_post_ref01"))
    render_page_post_ref01_data["api_key"] = setup[:idmap]["api_key01"]

    render_page_post_ref01_data_result = render_page_post_ref01_ent.create(render_page_post_ref01_data, nil)
    render_page_post_ref01_data = Helpers.to_map(render_page_post_ref01_data_result.respond_to?(:data_get) ? render_page_post_ref01_data_result.data_get : render_page_post_ref01_data_result)
    assert !render_page_post_ref01_data.nil?
    assert !render_page_post_ref01_data["id"].nil?

  end
end

def render_page_post_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "render_page_post", "RenderPagePostTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = PhantomjscloudSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["render_page_post01", "render_page_post02", "render_page_post03", "api_key01"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["PHANTOMJSCLOUD_TEST_RENDER_PAGE_POST_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "PHANTOMJSCLOUD_TEST_RENDER_PAGE_POST_ENTID" => idmap,
    "PHANTOMJSCLOUD_TEST_LIVE" => "FALSE",
    "PHANTOMJSCLOUD_TEST_EXPLAIN" => "FALSE",
    "PHANTOMJSCLOUD_APIKEY" => "NONE",
  })

  idmap_resolved = Helpers.to_map(
    env["PHANTOMJSCLOUD_TEST_RENDER_PAGE_POST_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["PHANTOMJSCLOUD_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
        "apikey" => env["PHANTOMJSCLOUD_APIKEY"],
      },
      extra || {},
    ])
    client = PhantomjscloudSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["PHANTOMJSCLOUD_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["PHANTOMJSCLOUD_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end
